platform/BufferedSerial.cpp

```diff
@@ -0,0 +1,258 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if DEVICE_SERIAL
+
+#include <errno.h>
+#include "platform/BufferedSerial.h"
+#include "platform/mbed_poll.h"
+#ifdef MBED_CONF_RTOS_PRESENT
+#include "rtos/rtos.h"
+#endif
+
+namespace mbed {
+
+BufferedSerial::BufferedSerial(PinName tx, PinName rx, int baud) :
+        SerialBase(tx, rx, baud),
+        _blocking(true),
+        _tx_irq_enabled(false),
+        _dcd(NULL)
+{
+    /* Attatch IRQ routines to the serial device. */
+    SerialBase::attach(callback(this, &BufferedSerial::rx_irq), RxIrq);
+}
+
+BufferedSerial::~BufferedSerial()
+{
+    delete _dcd;
+}
+
+void BufferedSerial::DCD_IRQ()
+{
+    _poll_change(this);
+}
+
+void BufferedSerial::set_data_carrier_detect(PinName DCD_pin, bool active_high)
+{
+    delete _dcd;
+    _dcd = NULL;
+
+    if (DCD_pin != NC) {
+        _dcd = new InterruptIn(DCD_pin);
+        if (active_high) {
+            _dcd->fall(callback(this, &BufferedSerial::DCD_IRQ));
+        } else {
+            _dcd->rise(callback(this, &BufferedSerial::DCD_IRQ));
+        }
+    }
+}
+
+int BufferedSerial::close()
+{
+    /* Does not let us pass a file descriptor. So how to close ?
+     * Also, does it make sense to close a device type file descriptor*/
+    return 0;
+}
+
+int BufferedSerial::isatty()
+{
+    return 1;
+
+}
+
+off_t BufferedSerial::seek(off_t offset, int whence)
+{
+    /*XXX lseek can be done theoratically, but is it sane to mark positions on a dynamically growing/shrinking
+     * buffer system (from an interrupt context) */
+    return -ESPIPE;
+}
+
+int BufferedSerial::sync()
+{
+    lock();
+
+    while (!_txbuf.empty()) {
+        unlock();
+#ifdef MBED_CONF_RTOS_PRESENT
+        // Doing better than yield would require TxIRQ to also do poll_change when becoming empty. Worth it?
+        rtos::Thread::yield();
```

Comment:
No, the wait will yield to other tasks in an RTOS environment.

And by stopping running this thread, it will also let lower-priority tasks run, which may be important.

I guess in the absence of a proper wake mechanism, wait_ms(1) is probably a bit nicer to other threads. Probably shouldn't impact the performance much - we may take 1ms to notice buffer changes, but then we'll have few bytes of data ready to process in one go.